From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Otto=20Pfl=C3=BCger?= <otto.pflueger@abscue.de>
Date: Tue, 15 Apr 2025 10:20:50 +0200
Subject: [PATCH 007/225] ASoC: sprd: do not keep reference to DMA parameters
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

The DMA parameters may be freed after hw_free is called, so it cannot
be assumed that the pointer will remain valid. Since the parameters are
not used outside hw_params, just pass them directly to
sprd_pcm_request_dma_channel instead of storing them as part of the
driver's private state.

Signed-off-by: Otto Pflüger <otto.pflueger@abscue.de>
---
 sound/soc/sprd/sprd-pcm-dma.c | 15 +++++----------
 1 file changed, 5 insertions(+), 10 deletions(-)

diff --git a/sound/soc/sprd/sprd-pcm-dma.c b/sound/soc/sprd/sprd-pcm-dma.c
index cbf5bf82d..9bd4ec70d 100644
--- a/sound/soc/sprd/sprd-pcm-dma.c
+++ b/sound/soc/sprd/sprd-pcm-dma.c
@@ -28,7 +28,6 @@ struct sprd_pcm_dma_data {
 
 struct sprd_pcm_dma_private {
 	struct snd_pcm_substream *substream;
-	struct sprd_pcm_dma_params *params;
 	struct sprd_pcm_dma_data data[SPRD_PCM_CHANNEL_MAX];
 	int hw_chan;
 	int dma_addr_offset;
@@ -155,12 +154,12 @@ static void sprd_pcm_release_dma_channel(struct snd_pcm_substream *substream)
 
 static int sprd_pcm_request_dma_channel(struct snd_soc_component *component,
 					struct snd_pcm_substream *substream,
+					struct sprd_pcm_dma_params *dma_params,
 					int channels)
 {
 	struct snd_pcm_runtime *runtime = substream->runtime;
 	struct sprd_pcm_dma_private *dma_private = runtime->private_data;
 	struct device *dev = component->dev;
-	struct sprd_pcm_dma_params *dma_params = dma_private->params;
 	int i;
 
 	if (channels > SPRD_PCM_CHANNEL_MAX) {
@@ -203,17 +202,13 @@ static int sprd_pcm_hw_params(struct snd_soc_component *component,
 	dma_params = snd_soc_dai_get_dma_data(snd_soc_rtd_to_cpu(rtd, 0), substream);
 	if (!dma_params) {
 		dev_warn(component->dev, "no dma parameters setting\n");
-		dma_private->params = NULL;
 		return 0;
 	}
 
-	if (!dma_private->params) {
-		dma_private->params = dma_params;
-		ret = sprd_pcm_request_dma_channel(component,
-						   substream, channels);
-		if (ret)
-			return ret;
-	}
+	ret = sprd_pcm_request_dma_channel(component, substream, dma_params,
+					   channels);
+	if (ret)
+		return ret;
 
 	sg_num = totsize / period;
 	dma_private->dma_addr_offset = totsize / channels;
